tests/qnodes/test_qnode_passthru.py

```diff
@@ -0,0 +1,186 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Unit tests for the :mod:`pennylane` :class:`.PassthruQNode` class.
+"""
+import pytest
+import numpy as np
+
+try:
+    import tensorflow as tf
+
+    if tf.__version__[0] == "1":
+        tf = None  # default.tensor.tf requires TF 2
+    else:
+        from tensorflow import Variable
+except ImportError:
+    tf = None
+
+import pennylane as qml
+from pennylane.qnodes import PassthruQNode, BaseQNode, JacobianQNode
+
+
+
+@pytest.fixture(scope="function")
+def mock_qnode(mock_device):
+    """Simple PassthruQNode with default properties."""
+
+    def circuit(x):
+        qml.RX(x, wires=0)
+        return qml.expval(qml.PauliZ(0))
+
+    node = PassthruQNode(circuit, mock_device)
+    return node
+
+
+class TestPassthruBasics:
+    """Tests basic PassthruQNode properties."""
+
+    def test_always_mutable(self, mock_qnode):
+        """PassthruQNodes are always mutable."""
+        assert mock_qnode.mutable
+
+    def test_repr(self, mock_qnode):
+        """String representation."""
+        assert repr(mock_qnode) == "<PassthruQNode: device='mock_device', func=circuit, wires=2>"
+
+
+@pytest.mark.skipif(tf is None, reason="TensorFlow 2.0 not found.")
+@pytest.fixture(scope="function")
+def tensornet_tf_device():
+    return qml.device('default.tensor.tf', wires=2)
+
+
+@pytest.mark.skipif(tf is None, reason="TensorFlow 2.0 not found.")
+class TestPassthruTF:
+    """Test that TF objects can be successfully passed through to a TF simulator device, and back to user."""
+
+    # real data type used by the default.tensor.tf plugin (TensorFlow is strict about types)
+    DTYPE = tf.float64 if tf else None
```

Comment:
Don't think the `if tf` is needed here (redundant)

Reply:
My understanding is that Python will evaluate the `if tf` part first. If `tf` is False, then `tf.float64` would produce an attribute error, but instead we return `None`

Reply:
Right.

Reply:
I meant more that this line will never be hit since pytest already skills execution when `tf is None`

Reply:
At first I thought so too, but the class is actually constructed first, which involves executing all statements inside it.

Reply:
Oh that's really good to know!

Reply:
I guess it makes sense, since the decorator is a function applied on the object it's decorating, in this case the class object, so the class needs to be constructed before it can be passed as an argument to the function.